File: c/isogram/src/isogram.c
#include "isogram.h"

bool is_isogram(const char *const str)
{
    int char_repetitions[26] = { 0 };
    
    for (size_t i = 0; str[i] != '\0'; i++)
    {
        if (!isalpha(str[i]))
        {
            continue;
        }
        
        int index;
        if (islower(str[i]))
        {
            index = str[i] - 'a';
        }
        else
        {
            index = str[i] - 'A';
        }
        
        if (++char_repetitions[index] > 1)
        {
            return false;
        }
    }
    
    return true;
}

File: javascript/isogram/isogram.js
class Isogram {
    constructor(str) {
        this._isIsogram = this._isIsogram(str);
    }

    isIsogram() {
        return this._isIsogram;
    }

    _isIsogram(str) {
        let charRepetitions = {};

        for (let i = 0; i < str.length; i++) {
            let c = str[i].toLowerCase();

            if (!this._isRelevantChar(c)) {
                continue;
            }

            if (charRepetitions[c] === undefined) {
                charRepetitions[c] = 1;
            }
            else {
                charRepetitions[c]++;
            }
        }

        return Object.keys(charRepetitions).filter(e => charRepetitions[e] > 1).length == 0;
    }

    _isRelevantChar(c) {
        return c !== ' ' && c !== '-';
    }
};

module.exports = Isogram;
